app/models/user.rb

```diff
@@ -0,0 +1,17 @@
+class User < ApplicationRecord
+  before_save :downcase_email
+    
+  validates :name,  presence: true, length: {maximum: Settings.username_maximum}
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
+  validates :email, presence: true, length: {maximum: Settings.email_maximum}, 
+  format: { with: VALID_EMAIL_REGEX },
+  uniqueness: true
+  has_secure_password
```

Comment:
cái này là cái gì

Reply:
has_secure_password là một phương thức trong Rails được sử dụng để mã hóa và xác thực mật khẩu bằng BCryt. Nó sẽ giả định trong cơ sở dữ liệu sẽ có một cột password_digest.
Phương thức trên với các chức năng:
+password_digest: mật khẩu được băm và trước khi lưu trữ an toàn trong cở sở dữ liệu
+Cặp thuộc tính password_confirmation: cần phải điền mật khẩu và xác nhận lại mật khẩu đúng với mật khẩu đã điền
+authenticate: sẽ trả về người dùng nếu ta điền đúng mật khẩu
Nếu không muốn xác thực, ta có thể dùng "validations: false", đồng thời ta cũng có thể tự thêm các yêu cầu xác thực khác.